horovod/tensorflow/__init__.py

```diff
@@ -188,6 +194,9 @@ def compute_gradients(self, *args, **kwargs):
             with tf.name_scope(self._name + "_Allreduce"):
                 for grad, var in gradients:
                     if grad is not None:
+                        if self._sparse_as_dense and \
```

Comment:
Would it make more sense to add this parameter to `allreduce` and do the casting there?  It would make the API more versatile, and reduce duplication across TensorFlow and Keras DistributedOptimizer implementations.

Reply:
I just feel weird about `hvd.allreduce()` turning sparse tensor into dense tensor, but I guess explicit `sparse_as_dense` flag would make it ok?

Reply:
Actually, the only thing it'd safe is these two lines:
```
if self._sparse_as_dense and isinstance(grad, tf.IndexedSlices):
   grad = tf.convert_to_tensor(grad)
```

We'd still need to have optimizers take `sparse_as_dense` flag, plumb it through and have docs.  So, until somebody asks for it, I'm still leaning towards making it optimizer-specific thing.

Reply:
I think that's fine.  I agree that it feels weird having `allreduce` turn a sparse tensor dense, but in fairness I also think it's weird that it will call `allgather` if the input is a `tf.IndexedSlices`.  But I agree, for now it's probably fine just to put it in the optimizer.